src/Stratis.Bitcoin/Consensus/PHConsensusManagerBehavior.cs

```diff
@@ -0,0 +1,306 @@
+                {
+                    var posBock = new PosBlock(result.Consumed.Header);
+                    ProvenBlockHeader provenBlockHeader = consensusFactory.CreateProvenBlockHeader(posBock);
+                    int consumedCount = headers.IndexOf(provenBlockHeader) + 1;
+                    this.cachedHeaders.AddRange(headers.Skip(consumedCount));
+
+                    this.logger.LogTrace("{0} out of {1} items were not consumed and added to cache.", headers.Count - consumedCount, headers.Count);
+                }
+
+                if (this.cachedHeaders.Count == 0)
+                    await this.ResyncAsync().ConfigureAwait(false);
+            }
+        }
+
+        /// <summary>Constructs the proven headers payload from locator to consensus tip.</summary>
+        /// <param name="locator">Block locator.</param>
+        /// <param name="hashStop">Hash of the block after which constructing headers payload should stop.</param>
+        /// <param name="lastHeader"><see cref="ProvenBlockHeader"/> of the last header that was added to the <see cref="ProvenHeadersPayload"/>.</param>
+        /// <returns><see cref="ProvenHeadersPayload"/> with headers from locator towards consensus tip or <c>null</c> in case locator was invalid.</returns>
+        private ProvenHeadersPayload ConstructHeadersPayload(BlockLocator locator, uint256 hashStop, out ProvenBlockHeader lastHeader)
+        {
+            ChainedHeader fork = this.chain.FindFork(locator);
+
+            lastHeader = null;
+
+            if (fork == null)
+            {
+                this.logger.LogTrace("(-)[INVALID_LOCATOR]:null");
+                return null;
+            }
+
+            var headers = new ProvenHeadersPayload();
+            var consensusFactory = new PosConsensusFactory();
+            foreach (ChainedHeader header in this.chain.EnumerateToTip(fork).Skip(1))
+            {
+                var posBock = new PosBlock(header.Header);
+                ProvenBlockHeader provenBlockHeader = consensusFactory.CreateProvenBlockHeader(posBock);
+                lastHeader = provenBlockHeader;
+                headers.Headers.Add(provenBlockHeader);
+
+                if ((header.HashBlock == hashStop) || (headers.Headers.Count == MaxItemsPerHeadersMessage))
+                    break;
+            }
+
+            return headers;
+        }
+
+        /// <summary>Validates the proven headers payload.</summary>
+        /// <param name="peer">The peer who sent the payload.</param>
+        /// <param name="provenHeadersPayload">Proven Headers payload to validate.</param>
+        /// <param name="validationError">The validation error that is set in case <c>false</c> is returned.</param>
+        /// <returns><c>true</c> if payload was valid, <c>false</c> otherwise.</returns>
+        private bool ValidateHeadersPayload(INetworkPeer peer, ProvenHeadersPayload provenHeadersPayload, out string validationError)
```

Comment:
looks like all logic is copypasted from base class. please reuse logic from there